File: flink-streaming/src/main/scala/edu/wzm/pvuv/PvUvExample.scala
package edu.wzm.pvuv

import org.apache.flink.api.common.functions.AggregateFunction
import org.apache.flink.api.java.utils.ParameterTool
import org.apache.flink.api.scala._
import org.apache.flink.streaming.api.TimeCharacteristic
import org.apache.flink.streaming.api.functions.source.SourceFunction
import org.apache.flink.streaming.api.functions.timestamps.BoundedOutOfOrdernessTimestampExtractor
import org.apache.flink.streaming.api.scala.StreamExecutionEnvironment
import org.apache.flink.streaming.api.windowing.assigners.TumblingEventTimeWindows
import org.apache.flink.streaming.api.windowing.time.Time
import org.apache.flink.streaming.api.windowing.triggers.CountTrigger

import java.util.{Set => JSet, HashSet => JHashSet}

/**
  * @author: Jimmy Wong
  * @Date: 2019-12-26
  * @version: 1.0
  * @Description: Example of PV and UV
  */
case class Event(date: String,
                 eventId: String,
                 timestamp: Long)

case class PvUvAccumulator(var date: String,
                           var pv: Long,
                           var uv: Long,
                           /**
                             * 这里需要用Java Collection，否则报错：
                             * org.apache.flink.shaded.guava18.com.google.common.util.concurrent.UncheckedExecutionException: java.lang.NumberFormatException: Not a version: 9
                             */
                           var ids: JSet[String])

case class PvUv(date: String,
                pv: Long,
                uv: Long)

object Driver {

    val SPACE_SEPARATOR = "\\s+"

    def main(args: Array[String]): Unit = {
        val params = ParameterTool.fromArgs(args)
        val env = StreamExecutionEnvironment.getExecutionEnvironment
        env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime)
        env.setParallelism(1)

        val stream = env.socketTextStream("localhost", 9999)

        stream.assignTimestampsAndWatermarks(new BoundedOutOfOrdernessTimestampExtractor[String](Time.seconds(10)) {
              override def extractTimestamp(msg: String): Long = {
                  msg.split(SPACE_SEPARATOR)(2).toLong
              }
          })
          .map(msg => {
              val info = msg.split(SPACE_SEPARATOR)
              Event(info(0), info(1), info(2).toLong)
          })
          .keyBy("date")
          .window(TumblingEventTimeWindows.of(Time.seconds(10)))
          .trigger(CountTrigger.of(1))
          .aggregate(new PvUvAggregator)
          .print()

        env.execute("PVUV")
    }
}

class PvUvAggregator extends AggregateFunction[Event, PvUvAccumulator, PvUv]{
    override def createAccumulator(): PvUvAccumulator = {
        PvUvAccumulator("", 0L, 0L, new JHashSet[String]())
    }

    override def add(value: Event, acc: PvUvAccumulator): PvUvAccumulator = {
        if (!acc.ids.contains(value.eventId)) {
            acc.ids.add(value.eventId)
            acc.uv = acc.uv + 1
            acc.date = value.date
        }
        acc.pv = acc.pv + 1

        acc
    }

    override def getResult(acc: PvUvAccumulator): PvUv = {
        PvUv(acc.date, acc.pv, acc.uv)
    }

    override def merge(acc1: PvUvAccumulator, acc2: PvUvAccumulator): PvUvAccumulator = {
        acc1.ids.addAll(acc2.ids)
        PvUvAccumulator(acc1.date, acc1.pv + acc2.pv, acc1.uv + acc2.uv, acc1.ids)
    }
}


class RandomFunction extends SourceFunction[(Long, Long)] {
    var flag = true

    override def cancel(): Unit = {
        flag = false
    }

    override def run(ctx: SourceFunction.SourceContext[(Long, Long)]): Unit = {
        while (flag) {
            val event = ((Math.random() * 1000).toLong, System.currentTimeMillis())
            ctx.collect(event)
            Thread.sleep(500)
        }
    }
}

File: flink-streaming/src/main/java/edu/wzm/pvuv/PvUvExample.java
package edu.wzm.pvuv;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.ToString;
import org.apache.flink.api.common.functions.AggregateFunction;
import org.apache.flink.api.java.utils.ParameterTool;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.functions.timestamps.BoundedOutOfOrdernessTimestampExtractor;
import org.apache.flink.streaming.api.windowing.assigners.TumblingProcessingTimeWindows;
import org.apache.flink.streaming.api.windowing.time.Time;
import org.apache.flink.streaming.api.windowing.triggers.CountTrigger;

import java.util.HashSet;
import java.util.Set;

/**
 * @author: Jimmy Wong
 * @Date: 2019-12-26
 * @version: 1.0
 * @Description: Example of PV and UV
 */
public class PvUvExample {

    private static final String SEPARATOR = "\\s+";

    public static void main(String[] args) throws Exception {
        // the port to connect to
        final int port;
        try {
            final ParameterTool params = ParameterTool.fromArgs(args);
            port = params.getInt("port");
        } catch (Exception e) {
            System.err.println("No port specified. Please run 'SocketWindowWordCount --port <port>'");
            return;
        }

        // get the execution environment
        final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();

        // get input data by connecting to the socket
        DataStream<String> text = env.socketTextStream("localhost", port, "\n");

        // parse the data, group it, window it, and aggregate the counts
        DataStream<PvUv> PvUvDataStream = text
            .assignTimestampsAndWatermarks(new BoundedOutOfOrdernessTimestampExtractor<String>(Time.seconds(10)) {
                @Override
                public long extractTimestamp(String s) {
                    return Long.parseLong(s.split(SEPARATOR)[2]);
                }
            })
            .map(value -> {
                String[] vals = value.split(SEPARATOR);
                return new Event(vals[0], vals[1], Long.parseLong(vals[2]));
            })
            .keyBy("date")
            .window(TumblingProcessingTimeWindows.of(Time.seconds(10)))
            .trigger(CountTrigger.of(1))
            .aggregate(new PvUvAggregator())
            .map(acc -> PvUv.of(acc.getDate(), acc.getPv(), acc.getUv()));

        // print the results with a single thread, rather than in parallel
        PvUvDataStream.print().setParallelism(1);

        env.execute("Socket Window WordCount");
    }

    private static class PvUvAggregator implements AggregateFunction<Event, PvUvAccumulator, PvUv> {
        @Override
        public PvUvAccumulator createAccumulator() {
            return PvUvAccumulator.of(0L, 0L);
        }

        @Override
        public PvUvAccumulator add(Event value, PvUvAccumulator accumulator) {
            if (!accumulator.getIds().contains(value.getEventId())) {
                accumulator.getIds().add(value.getEventId());
                accumulator.setUv(accumulator.getUv() + 1);
                accumulator.setDate(value.getDate());
            }

            accumulator.setPv(accumulator.getPv() + 1);
            return accumulator;
        }

        @Override
        public PvUv getResult(PvUvAccumulator accumulator) {
            return PvUv.of(accumulator.getDate(), accumulator.getPv(), accumulator.getUv());
        }

        @Override
        public PvUvAccumulator merge(PvUvAccumulator acc1, PvUvAccumulator acc2) {
            acc1.getIds().addAll(acc2.getIds());
            return PvUvAccumulator.of(acc1.getDate(),
                    acc1.getPv() + acc2.getPv(),
                    acc1.getUv() + acc2.getUv(),
                    acc1.getIds());
        }
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @ToString
    public static class Event {
        private String date;
        private String eventId;
        private long timestamp;
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @ToString
    public static class PvUv {
        private String date;
        private long pv;
        private long uv;

        public static PvUv of(String date, long pv, long uv) {
            return new PvUv(date, pv, uv);
        }
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @ToString
    public static class PvUvAccumulator {
        private String date;
        private long pv;
        private long uv;
        private Set<String> ids;

        public static PvUvAccumulator of(long pv, long uv) {
            return new PvUvAccumulator("", pv, uv, new HashSet<>());
        }

        public static PvUvAccumulator of(String date, long pv, long uv, Set<String> eventIds) {
            return new PvUvAccumulator(date, pv, uv, eventIds);
        }
    }
}